presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java

```diff
@@ -968,4 +969,26 @@ public static void deleteDirectoryRecursively(HdfsContext context, HdfsEnvironme
             log.warn(e, "Failed to delete path: " + path.toString());
         }
     }
+
+    public static Map<String, String> updateMetadataWithPreviousPartitionVersion(Map<String, String> extraPartitionMetadata, Optional<Partition> existingPartition)
+    {
+        Optional<Long> existingPartitionVersion = existingPartition.map(Partition::getPartitionVersion).orElse(Optional.empty());
+        if (!existingPartitionVersion.isPresent()) {
+            return extraPartitionMetadata;
+        }
+        return ImmutableMap.<String, String>builder()
+                .putAll(extraPartitionMetadata)
+                .put(PARTITION_VERSION, String.valueOf(existingPartitionVersion.get()))
+                .build();
+    }
+
+    public static Optional<Long> toMetastorePartitionVersion(String partitionVersion)
+    {
+        try {
+            return Optional.of(Long.parseLong(partitionVersion));
+        }
+        catch (NumberFormatException exception) {
+            return Optional.empty();
+        }
+    }
```

Comment:
Instead of adding partition version to parameters, how about we add `Optional<Long> partitionVersion` as additional parameter to `PartitionObjectBuilder.buildPartitionObject` ? and then we set the partition version only in `PrismPartitionObjectBuilder` implementation

Reply:
Yeah makes sense to use the version explicitly.